flink-connector-prometheus/src/main/java/org/apache/flink/connector/prometheus/table/PrometheusDynamicSinkFactory.java

```diff
@@ -0,0 +1,165 @@
+                                        flushBufferSize));
+        Optional.ofNullable(config.get(AsyncSinkConnectorOptions.MAX_BATCH_SIZE))
+                .ifPresent(
+                        maxBatchSize ->
+                                properties.put(
+                                        AsyncSinkConnectorOptions.MAX_BATCH_SIZE.key(),
+                                        maxBatchSize));
+        Optional.ofNullable(config.get(AsyncSinkConnectorOptions.MAX_IN_FLIGHT_REQUESTS))
+                .ifPresent(
+                        maxInflightRequests ->
+                                properties.put(
+                                        AsyncSinkConnectorOptions.MAX_IN_FLIGHT_REQUESTS.key(),
+                                        maxInflightRequests));
+        Optional.ofNullable(config.get(AsyncSinkConnectorOptions.MAX_BUFFERED_REQUESTS))
+                .ifPresent(
+                        maxBufferedRequests ->
+                                properties.put(
+                                        AsyncSinkConnectorOptions.MAX_BUFFERED_REQUESTS.key(),
+                                        maxBufferedRequests));
+        Optional.ofNullable(config.get(AsyncSinkConnectorOptions.FLUSH_BUFFER_TIMEOUT))
+                .ifPresent(
+                        timeout ->
+                                properties.put(
+                                        AsyncSinkConnectorOptions.FLUSH_BUFFER_TIMEOUT.key(),
+                                        timeout));
+        return properties;
+    }
+
+    private void validateMetricConfigKeys(
+            PrometheusConfig prometheusConfig, DataType physicalDataType) {
+        Set<String> fieldNames =
+                DataType.getFields(physicalDataType).stream()
+                        .map(DataTypes.AbstractField::getName)
+                        .collect(Collectors.toSet());
+
+        List<String> metricConfigKeys =
+                Stream.concat(
+                                Stream.of(
+                                        prometheusConfig.getMetricName(),
+                                        prometheusConfig.getMetricSampleKey(),
+                                        prometheusConfig.getMetricSampleTimestamp()),
+                                prometheusConfig.getLabelKeys().stream())
+                        .collect(Collectors.toList());
+
+        metricConfigKeys.forEach(
+                metricConfigKey ->
+                        Optional.ofNullable(metricConfigKey)
+                                .filter(fieldNames::contains)
+                                .orElseThrow(
+                                        () ->
+                                                new ValidationException(
+                                                        String.format(
+                                                                "%s not found in fields: %s",
```

Comment:
Could we consider something along the lines of `%s is not a valid Prometheus metric config key, valid keys are: %s`, this message is exposing the implementation of the validation rather than what the user should do to fix it.

Reply:
Adopted suggestion